Eventing updates for Release/7.1 (#2696)
couchbase/docs-server: modules/eventing/pages/eventing-rbac.adoc

Couchbase Server Enterprise Edition provides RBAC with multiple roles for finer access control.
Community Edition provides multiple users that can be assigned to limited set of roles.
There are three fixed roles in the community edition of Couchbase providing coarser access control: Bucket Full Access (`bucket_full_access[*]`), Admin (`admin`), and Read Only Admin (`ro_admin`).

A Couchbase-Server _role_ permits one or more _resources_ to be accessed according to defined _privileges_.
Roles can be assigned to individual users, and to groups, by means of either the UI or the REST API.

For a complete list of roles, see xref:learn:security/roles.adoc[Roles].
Note that most roles can be assigned only on the _Enterprise Edition_ of Couchbase Server: on the _Community Edition_ of Couchbase Server, only the `bucket_full_access`,
`admin`, and `ro_admin` roles can be assigned.

For more information, see xref:learn:security/authorization-overview.adoc[Authorization], and xref:manage:manage-security/manage-users-and-roles.adoc[Manage Users, Groups, and Roles].

== Function Scope

A bucket.scope combination is used for identifying functions belonging to the same group.

> Only the "Eventing Full Admin" role and also the "Full Admin" role can set the bucket.scope to  *+*+.+*+* all other Eventing non-privileged users need to define a *Function Scope* for their Eventing functions that references an existing resource of bucket.scope. This provides role based isolation of Eventing functions between non-privileged users

Reviewer: ```suggestion
Only the "Eventing Full Admin" role and also the "Full Admin" role can set the bucket.scope to  *+*+.+*+*; all other Eventing non-privileged users need to define a *Function Scope* for their Eventing functions that references an existing resource of bucket.scope. 
This provides role based isolation of Eventing functions between non-privileged users
```

> Typically you should set Function Scope to the bucket.scope that holds the collection that is the source of your mutations to your Eventing Function.  This best practice ensures that you _*do not*_  inadvertently cause an Eventing Function to undeploy by removing a *Function Scope* pointing to a resource that is not required for the function to run.

Reviewer: ```suggestion
Typically you should set Function Scope to the bucket.scope that holds the collection that is the source of your mutations to your Eventing Function.  This best practice ensures that you _do not_  inadvertently cause an Eventing Function to undeploy by removing a *Function Scope* pointing to a resource that is not required for the function to run.
```

Reviewer: Just checking. Should this be:

This best practice ensures that you _do not_  inadvertently cause an Eventing Function to undeploy by removing a *Function Scope* pointing to a resource that is required for the function to run.

== Privileged Users

> If a user role of either "Full Admin" or "Eventing Full Admin" it by default has all the necessary access to every resource in a cluster required to run the Eventing Service and create and manage Eventing Functions.

Reviewer: ```suggestion
If a user role of either "Full Admin" or "Eventing Full Admin", then by default this user has all the necessary access to every resource in a cluster required to run the Eventing Service and create and manage Eventing Functions.
```

> When creating and Eventing Function either of these roles can set the *Function Scope* to *+*+.+*+* no other RBAC role is allowed to use this *Function Scope*.

Reviewer: ```suggestion
When creating and Eventing Function either of these roles can set the *Function Scope* to *+*+.+*+*;  no other RBAC role is allowed to use this *Function Scope*.
```

NOTE: When upgrading to 7.1 all Eventing Functions are assumed to be running as a privileged user and have their *Function Scope* set to *+*+.+*+* to ensure continuity of your Eventing Functions.

=== Full Admin v. Eventing Full Admin

> Prior to 7.0.0 Eventing always run as "Full Admin" this blocked some use cases and adoption as this role allowed creation of new users and the ability to escalate privilege sets. The  "Eventing Full Admin" role introduced in 7.0 simply removes the capability of creating users and assigning/modifying RBAC credentials thus providing a bit more security.

Reviewer: ```suggestion
Prior to 7.0.0 Eventing always run as "Full Admin". This blocked some use cases and adoption as this role allowed creation of new users and the ability to escalate privilege sets. The  "Eventing Full Admin" role introduced in 7.0 simply removes the capability of creating users and assigning/modifying RBAC credentials thus providing a bit more security.
```

> For the Function Scope or RBAC grouping we will use the 'bulk.data' assuming you have the role of either "Full Admin" or "Eventing Full Admin". For standard or non-privileged users refer to Role-Based Access Control.

Reviewer: ```suggestion
For the Function Scope or RBAC grouping, we will use the 'bulk.data' assuming you have the role of either "Full Admin" or "Eventing Full Admin". 
For standard or non-privileged users refer to Role-Based Access Control.
```

== Eventing and RBAC for Non-privileged Users

> NOTE: If a user role of either "Full Admin" or "Eventing Full Admin" it by default has all the necessary access to every resource in a cluster required to run the Eventing Service and create and manage Eventing Functions.

Reviewer: ```suggestion
NOTE: If a user role of either "Full Admin" or "Eventing Full Admin", then this user, by default, has all the necessary access to every resource in a cluster required to run the Eventing Service and create and manage Eventing Functions.
```

> _In RBAC although you can assign rolls directly to a *USER* it is generally more flexible if you define a *GROUP* and assign that group or set of roles to a *USER*, this allows reusing a *GROUP* across multiple users._

Reviewer: ```suggestion
_In RBAC although you can assign rolls directly to a *USER*,  it is generally more flexible if you define a *GROUP* and assign that group or set of roles to a *USER*, this allows reusing a *GROUP* across multiple users._
```

== Minimal Eventing RBAC role

The following minimal resources are required for a non-privileged user to access the Eventing Service and create and manage Eventing Functions.

* Data Reader
** Eventing Storage keyspace or Scratchpad
* Data Writer
** Eventing Storage keyspace or Scratchpad
* Data DCP reader
** Mutation Source
* Eventing / Manage Scope Function
** bucket.scope or bucket.*

== Minimal Eventing RBAC role example

* Access the Couchbase Web Console > Security and Select "ADD GROUP".
+
image::rbac_min_add_add_group.png[,%100]

> * Configure the group as follows "Data Reader" and "Data Writer" are required for the Eventing Storage or scratchpad. "Data DCP Reader" is required to fetch the mutations form DCP. Note this item was defined as 'bulk.data.*' which would allow building Evening functions that can listen to any collection under 'bulk.data'.

Reviewer: ```suggestion
* Configure the group as follows: "Data Reader" and "Data Writer" are required for the Eventing Storage or scratchpad. "Data DCP Reader" is required to fetch the mutations form DCP. Note this item was defined as 'bulk.data.*' which would allow building Evening functions that can listen to any collection under 'bulk.data'.
```

+
image::rbac_min_a.png[,538,align=middle]
+

> The final item required is defining the *Function Scope* under "Eventing / Manage Scope Function" since we will be listing to mutations in a collection under 'bulk.data' it makes sense to use this as our grouping.

Reviewer: ```suggestion
The final item required is defining the *Function Scope* under "Eventing / Manage Scope Function". Since we will be listing to mutations in a collection under 'bulk.data' it makes sense to use this as our grouping.
```

* Hit *Save* to store the USER to the system.

* Access the Couchbase Web Console > Security 

* Select GROUPS on the right, you should see your definition for GROUP "eventing_min"
+
image::rbac_min_groups.png[,%100]

* Select USERS on the right, you should see your definition for USER "user_min"
+
image::rbac_min_users.png[,%100]

== Beyond a Minimal Eventing RBAC role

You may consider adding

* Data Reader
** Mutation Source
* Data Writer
** Mutation Source
* Data Monitor
** Mutation Source
** Eventing Storage keyspace or Scratchpad

If you have any Bindings in your Eventing Function of type "Bucket Alias" you will need to have one or more additional settings if not already allowed.

* Data Reader
** Bucket Alias
* Data Writer
** Bucket Alias

If you plan to use N1QL consider adding at lease SELECT privileges

* Query & Index / Query Select
** Mutation Source

== Multi-tenancy in Eventing

The "Function Scope" in an Eventing Function works with the RBAC selection in "Eventing / Manage Scope Function" to limit access to between tenants in both the UI and the REST API. 

A tenant might be based on company departments such as administration, sales, production and support.

> Below we have two tenants example (an admin and a limited user) and four Eventing Functions each with a different *Function Scope*, we logged into the UI with either an Eventing Full Admin" or "Full Admin" role and thus we can access all of the  Eventing Functions. 

Reviewer: ```suggestion
Below we have two tenants example (an admin and a limited user) and four Eventing Functions each with a different *Function Scope*. 
We logged into the UI with either an Eventing Full Admin" or "Full Admin" role and thus we can access all of the  Eventing Functions. 
```

image::rbac_admin_view.png[,%100]

> Now log out of the UI console and log back in as a non-privileged user (for example we use the USER "user_min" as defined above).  Because of the privliges defined we are only allowed access to Eventing Functions that have a *Function Scope* of 'bulk.data'.

Reviewer: ```suggestion
Now log out of the UI console and log back in as a non-privileged user (for example we use the USER "user_min" as defined above).  
Because of the privliges defined we are only allowed access to Eventing Functions that have a *Function Scope* of 'bulk.data'.
```
